client/state/db_test.go

```diff
@@ -137,6 +138,102 @@ func TestStateDB_Allocations(t *testing.T) {
 	})
 }
 
+// TestStateDB_Batch asserts the behavior of PutAllocation, PutNetworkStatus and
+// DeleteAllocationBucket in batch mode, for all operational StateDB implementations.
+func TestStateDB_Batch(t *testing.T) {
+	t.Parallel()
+
+	testDB(t, func(t *testing.T, db StateDB) {
+		require := require.New(t)
+
+		// For BoltDB, get initial tx_id
+		var getTxID func() int
+		var prevTxID int
+		if boltStateDB, ok := db.(*BoltStateDB); ok {
+			boltdb := boltStateDB.DB().BoltDB()
+			getTxID = func() int {
+				tx, err := boltdb.Begin(true)
+				require.NoError(err)
+				defer tx.Rollback()
+				return tx.ID()
+			}
+			prevTxID = getTxID()
+		}
+
+		// Write 1000 allocations and network statuses in batch mode
+		var allocs []*structs.Allocation
+		for i := 0; i < 1000; i++ {
+			allocs = append(allocs, mock.Alloc())
+		}
+		var wg sync.WaitGroup
+		for _, alloc := range allocs {
+			wg.Add(1)
+			go func(alloc *structs.Allocation) {
+				require.NoError(db.PutNetworkStatus(alloc.ID, mock.AllocNetworkStatus(), WithBatchMode()))
+				require.NoError(db.PutAllocation(alloc, WithBatchMode()))
+				wg.Done()
+			}(alloc)
+		}
+		wg.Wait()
+
+		// Check BoltDB actually combined PutAllocation calls into much fewer transactions.
+		// The actual number of transactions depends on how fast the goroutines are spawned,
+		// with every 10ms period saved in a separate transaction (see boltdb MaxBatchDelay
+		// and MaxBatchSize parameters).
+		if getTxID != nil {
+			numTransactions := getTxID() - prevTxID
+			require.Less(numTransactions, 10)
```

Comment:
Was this "< 10 transactions" determined empirically? The default [MaxBatchSize](https://godoc.org/github.com/boltdb/bolt#pkg-constants) is 1000, which implies we can assert "at least 2" but I'm less certain about the "at most" we can assert. Can we use a timer to figure out how many 10ms periods have passed to make a stronger assertion?

I just want to make sure we're not introducing a potential test flake (this has been a [pain point](https://github.com/hashicorp/nomad/issues?q=is%3Aissue+flaky)), especially for full unit test runs with parallelism on the VMs used for CI.

Reply:
Yep flake tests is a common pain point - I've wasted so much time on it too.

You're right, most of the time when running this test we get 2 transactions, each with 1000 writes. Time is usually not a limiting factor here – spawning goroutines is pretty fast and writes are all in-memory (the only I/O we have is when these 2 transactions are committed). I used 10 as a "safe" threshold that should hold even under load, but that's of course arbitrary. 

I initially opted against adding a more specific "expected number of transactions" calculation here because that would add coupling to specific algorithm that boltdb uses, in addition to adding complexity. Now that I'm thinking about, the first concern is probably not relevant as boltdb is frozen, so ok, let's do it.

Reply:
LGTM